pygen/pygen_src/riscv_reg.py

```diff
@@ -0,0 +1,131 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+"""
+
+import sys
+import logging
+import vsc
+from importlib import import_module
+from pygen_src.riscv_instr_gen_config import cfg
+from pygen_src.riscv_instr_pkg import (
+    pkg_ins, privileged_reg_t, reg_field_access_t, privileged_level_t)
+rcs = import_module("pygen_src.target." + cfg.argv.target + ".riscv_core_setting")
+
+
+# Light weight RISC-V register class library
+# Base class for RISC-V register field
+@vsc.randobj
+class riscv_reg_field:
+    def __init__(self):
+        self.bit_width = 0
+        self.reset_val = vsc.bit_t(rcs.XLEN)
+        self.val = vsc.rand_bit_t(rcs.XLEN)
+        self.access_type = vsc.enum_t(reg_field_access_t)
+        self.hard_wired = vsc.bit_t(1)
+        self.name = ""
+
+    @vsc.constraint
+    def zero_reserved_field_c(self):
+        with vsc.implies(self.access_type == reg_field_access_t.WPRI):
+            self.val == 0
+
+    @vsc.constraint
+    def hardwired_fld_c(self):
+        with vsc.implies(self.hard_wired == 1):
+            self.val == self.reset_val
+
+    def covert2string(self):
+        return pkg_ins.format_string("{} bit_width:{} val:{} type:{}".format(
+            self, self.bit_width, hex(self.val), self.access_type))
+
+    def post_randomize(self):
+        mask = vsc.bit_t(rcs.XLEN)
+        mask = 1
+        mask = mask >> (rcs.XLEN - self.bit_width)
+        self.val = mask & self.val
+
+
+# Base class for RISC-V register
+@vsc.randobj
+class riscv_reg:
+    def __init__(self):
+        self.reg_name = vsc.enum_t(privileged_reg_t)
+        self.offset = vsc.bit_t(12)
+        self.privil_level = vsc.enum_t(privileged_level_t)
+        self.val = vsc.bit_t(rcs.XLEN)
+        self.fld = vsc.rand_list_t(vsc.attr(riscv_reg_field()))
+
+    def init_reg(self, reg_name):
+        self.reg_name = reg_name
+        self.offset = reg_name.value
+
+    def get_val(self):
+        total_len = 0
+        for i in range((len(self.fld) - 1), -1, -1):
+            total_len += self.fld[i].bit_width
+        if total_len != rcs.XLEN:
+            for i in range((len(self.fld) - 1), -1, -1):
+                logging.info(self.fld[i].convert2string())
+            logging.critical("Total field length {} != XLEN {}".format(total_len, rcs.XLEN))
+            sys.exit(1)
+        self.val = 0
+        for i in range((len(self.fld) - 1), -1, -1):
+            self.val = (self.val << self.fld[i].bit_width) | self.fld[i].val
+        return self.val
+
+    def add_field(self, fld_name, bit_width, access_type, reset_val = 0):
+        new_fld = fld_name
```

Comment:
I think this line is redundant and should be removed (as new_fld will be updated in the next line, and the fld_name will be assigned to the name field later on anyways).

Reply:
Correct!
Thanks, @hodjat91